chia/full_node/mempool.py

```diff
@@ -21,65 +29,139 @@ class MempoolRemoveReason(Enum):
     POOL_FULL = 3
 
 
+@dataclass(frozen=True)
+class InternalMempoolItem:
+    spend_bundle: SpendBundle
+    npc_result: NPCResult
+    height_added_to_mempool: uint32
+
+
 class Mempool:
+    _db_conn: sqlite3.Connection
+    # it's expensive to serialize and deserialize G2Element, so we keep those in
+    # this separate dictionary
+    _items: Dict[bytes32, InternalMempoolItem]
+
     def __init__(self, mempool_info: MempoolInfo, fee_estimator: FeeEstimatorInterface):
-        self._spends: Dict[bytes32, MempoolItem] = {}
-        self._sorted_spends: SortedDict = SortedDict()
+        self._db_conn = sqlite3.connect(":memory:")
+        self._items = {}
+
+        with self._db_conn:
+            # name means SpendBundle hash
+            # assert_height may be NIL
+            self._db_conn.execute(
+                """CREATE TABLE tx(
+                name BLOB PRIMARY KEY,
+                cost INT NOT NULL,
+                fee INT NOT NULL,
+                assert_height INT,
+                fee_per_cost REAL GENERATED ALWAYS AS (fee / cost) VIRTUAL)
+                """
+            )
+            self._db_conn.execute("CREATE INDEX fee_sum ON tx(fee)")
+            self._db_conn.execute("CREATE INDEX cost_sum ON tx(cost)")
+            self._db_conn.execute("CREATE INDEX feerate ON tx(fee_per_cost)")
```

Comment:
you could consider adding this index to support `spends_by_feerate`:
```
CREATE INDEX feerate_desc ON tx(fee_per_cost desc);
```

My benchmarks didn't bring a big improvement, but still the time for spends_by_feerate was reduced from ~24.5% to 24% on my testing machine.

Reply:
interesting. I assumed indices are ordered, without the traversal order mattering.